### ISSUE TITLE
kvserver: prioritize replication activity 

### ISSUE THREAD

**Issue by lunevalex** [cockroachdb/cockroach#81935]
In #79249 we observed that decommissioning can starve out load based rebalancing in a cluster. This led to a discussion that in certain customer workloads load based rebalancing is no longer a performance optimization, but a fundamental requirement for system stability. If a load-based lease transfer or replica movement is delayed or starved by any other replication activity then cluster can enter a degraded state and had an adverse effect on the workload. This line of reasoning led to the conclusion that there should be a natural priority order to rebalancing activity within a cluster. The replicate_queue attempts to do this when it picks the action to perform here https://github.com/cockroachdb/cockroach/blob/e4c7ca59b5d1858dc40ae31f7aac6a35fe09a777/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L650, but this is not enough as this decision is made on the individual range level. We should consider broader prioritization scheme at the store level that will prioritize snapshot/replication activity for the whole store. 

A naive prioritization order is 
- Snapshots to catchup existing followers 
- Load Based Replica Movement
- Upreplication
- Decommissioning
- Voter Rebalance
- All non-voter operations

There are many pitfalls to adding a prioritization scheme to replication/snapshot activity, as the flip-side of prioritization is starvation and so that will need to be considered here. How do we prioritize work without causing starvation? 

Jira issue: CRDB-16130

Epic CRDB-14621

**Comment by andrewbaptist** [comment]
*Overview*

The underlying issue is that during heavy snapshot transfers, other types of snapshots can be starved because of the way the receive queue is structured. 

There are really three separate issues to address
1) Prioritization on the send side
2) Fairness on the receive side
3) Limiting the performance impact of snapshot traffic

This story is focused on the second problem, fairness on the receive side. There are already stories for work on the sender side [84453](https://github.com/cockroachdb/cockroach/issues/84453) and others that [kvoli](https://github.com/kvoli) is working on. And there is separate work on the limiting impact side that [irfansharif]https://github.com/irfansharif is also working on. Neither of those is a candidate for v22.2, however, there is a small change on fairness that will address a majority of the issue. 

*Problem description*
In practice in a distributed system without central coordination, perfect prioritization is impossible because different stores have different views of the world. That said,  it can be greatly improved from the current situation. There have been two specific "combinations of snapshot transfers" that are important to look at, even though these are just representative sample.

- Decommissioning + Load-based rebalancing. Currently, if there is decommissioning work, load-based rebalancing can be starved. 
- Decommissioning + Range-based rebalancing. Currently, if there is range-based rebalancing, for example, if a store is added just before or during decommissioning, then it will "fight" for resources with the decommissioning work and potentially make it take much longer.

Ideally, this type of prioritization could be taken care of on the sender side, but in practice, this does not work well. What happens is that some senders will have no high priority work, and will shift to lower priority work, while other stores are still doing the high priority work. From the receiver today, it is impossible to tell the priority of the work when it is received. The examples I gave above are illustrative because both load-based rebalancing and range-based rebalancing are both viewed the same on the receiver side, but differently on the sender side.

Because of an unfortunate timeout limitation, it is not just that the low priority work delays the high priority work, it can completely starve the high priority work. The high priority work will continually hit a timeout and fail and retry, and make almost no progress until the low priority work is complete. We have seen this happen in numerous customer issues.

Separate there was a separate issue where on a single range, all actions for that range are processed before it moves onto the next range, even if they are lower priority. There is a [PR](https://github.com/cockroachdb/cockroach/pull/85219) out for that, but in practice, this is better, but insufficient because of the multiple store sender problem. It did improve the situation, but not enough. What happened in practice is that as each store finished decommissioning work, it shifted to range-based rebalance work and eventually 7/8 of the work being done was rebalancing, and the last few replicas took too long to get off the decommission store. 

*Proposed Solution*

The proposed solution is to have a set of queues for each separate queues for each type of work on the receive side
* Snapshots to catchup slow followers 
* Load-Based Replica Movement
* Add voters with N missing (multiple queues for different values of N)
* Decommissioning
* Range-Based rebalance
* Replica Constraint rebalance
* Pre-Merge alignment rebalance
* Add Non-Voter 
* Split and scatter after import
* User initiated scatter
* User initiated 
* User initiated range move
* User initiated replica move

There is no priority between these on the receiver side, but the receiver is able to see all the types of snapshots and put them in separate queues. As this list is likely both incomplete and will change over releases (for instance, we don't know the max N for N missing replicas), this is sent as a string rather than an Enum. On the receiver side, it puts the replicas into different FIFO queues and round-robins through the queues processing any non-empty queues. The assumption is that most of the time there will be only empty queues, but when the system is disrupted (user or system initiated) - then there will be multiple snapshots of different types and the goal is to be fair.

There is a map on the receive side `map(tag) -> channel`. Currently, when a request comes in, there is a code that looks like:

```
select {
    case snapshotSem <- struct{}{}:
        startSending()
}
```

This will change to 

```
mySemaphore = snapshotQueue.add(tag)
select {
    case mySemaphore <- struct{}{}:
        startSending()
}
```

The `snapshotQueue` will internally manage the set of tags and the round-robin behavior, allowing at most the `COCKROACH_CONCURRENT_SNAPSHOT_APPLY_LIMIT` (currently 1) number of snapshots to transfer in parallel. 

This proposal allows flexible addition of tags across versions, and the receiver doesn’t need to know the complete set of tags for any version. A current (v22.1) receiver that does not know about tags will just process things like it does today and will be no worse.

It might be possible to try and reuse the current `SnapshotRequest.Priority` field, but it is cleaner to add a new field `SnapshotRequest.Tag` and longer-term deprecate the existing one.


*Possible extensions*
    It is possible to allow receive side prioritization as well using this tag. Since the tags are not all known and since there is already sender-side prioritization, this may not be necessary but would be relatively simple to accomplish. A simple approach would be to have a runtime configurable map of "tag -> priority". The default priority could be 1.0, and the receiver could sort the queues by priority. It would then fully drain a high-priority queue before moving on to the next priority. It might make sense to add this option in the future once we better prioritize on the sender side, because having two prioritization schemes on the same task is complex to reason about. However, having this mechanism but not using it normally, may mean that in a future customer case, we could temporarily adjust the priority of only one queue to get us out of the situation (either making that queue higher or lower).
